com.unity.ml-agents/Runtime/Sensors/ObservationSpec.cs

```diff
@@ -0,0 +1,136 @@
+using Unity.Barracuda;
+
+namespace Unity.MLAgents.Sensors
+{
+    /// <summary>
+    /// A description of the observations that an ISensor produces.
+    /// This includes the size of the observation, the properties of each dimension, and how the observation
+    /// should be used for training.
+    /// </summary>
+    public struct ObservationSpec
+    {
+        internal readonly InplaceArray<int> m_Shape;
+
+        /// <summary>
+        /// The size of the observations that will be generated.
+        /// For example, a sensor that observes the velocity of a rigid body (in 3D) would use [3].
+        /// A sensor that returns an RGB image would use [Height, Width, 3].
+        /// </summary>
+        public InplaceArray<int> Shape
+        {
+            get => m_Shape;
+        }
+
+        internal readonly InplaceArray<DimensionProperty> m_DimensionProperties;
+
+        /// <summary>
+        /// The properties of each dimensions of the observation.
+        /// The length of the array must be equal to the rank of the observation tensor.
+        /// </summary>
+        /// <remarks>
+        /// It is generally recommended to use default values provided by helper functions,
+        /// as not all combinations of DimensionProperty may be supported by the trainer.
+        /// </remarks>
+        public InplaceArray<DimensionProperty> DimensionProperties
+        {
+            get => m_DimensionProperties;
+        }
+
+        internal ObservationType m_ObservationType;
+
+        /// <summary>
+        /// The type of the observation, e.g. whether they are generic or
+        /// help determine the goal for the Agent.
+        /// </summary>
+        public ObservationType ObservationType
+        {
+            get => m_ObservationType;
+        }
+
+        /// <summary>
+        /// The number of dimensions of the observation.
+        /// </summary>
+        public int Rank
+        {
+            get { return Shape.Length; }
+        }
+
+        /// <summary>
+        /// Construct an ObservationSpec for 1-D observations of the requested length.
+        /// </summary>
+        /// <param name="length"></param>
+        /// <returns></returns>
+        public static ObservationSpec Vector(int length)
```

Comment:
For Vector and Visual, add ObservationType to the static constructors.

Reply:
but make them default to Default? Sure, that sounds reasonable.

Reply:
@vincentpierre Any reason to not add that to VariableLength also?

Reply:
The goal conditioning trainer will not work with VariableLength obs. We can add it though, but the trainer will raise an error or a warning when this happens.

Reply:
Good call, removed the variable length.